lib/forspell/reporter.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'pastel'
+require 'logger'
+require 'json'
+
+module Forspell
+  class Reporter
+    SUCCESS_CODE = 0
+    ERROR_CODE = 1
+    FORMAT = '%<file>s:%<line>i: %<text>s (suggestions: %<suggestions>s)'
+
+    def initialize(logfile:,
+                   verbose:,
+                   format:,
+                   group:)
+
+      FileUtils.touch(logfile) if logfile.is_a?(String)
+      @logger = Logger.new(logfile || STDOUT)
+      @logger.level = verbose ? Logger::INFO : Logger::WARN
+
+      @logger.formatter = proc do |_severity, _datetime, _progname, msg|
+        "#{msg}\n"
+      end
+
+      @format = group ? 'group' : format
+
+      @pastel = Pastel.new(enabled: $stdout.tty?)
+      @errors = []
+      @files = []
+    end
+
+    def file(path)
+      @logger.info "Processing #{path}"
+      @files << path
+    end
+
+    def error(word, suggestions)
+      @errors << [word, suggestions]
+      puts readable(word, suggestions) if @format == 'readable'
+    end
+
+    def parsing_error(error)
+      @logger.error "PARSING ERROR IN #{@files.last}: #{error}"
+    end
+
+    def report
+      case @format
+      when 'readable'
+        err_count = @errors.size
+
+        color = err_count.positive? ? :red : :green
+        total_errors_colorized = @pastel.decorate(err_count.to_s, color)
+
+        puts 'Forspell inspects *.rb, *.c, *.cpp, *.md files'
+        puts "#{@files.size} files inspected, #{total_errors_colorized} errors detected"
+      when 'group'
+        @errors.flat_map(&:first)
+               .group_by { |word| word[:text] }
+               .transform_values { |v| v.map { |word| word[:file] }.uniq }
```

Comment:
`word` is `Struct` here, right?.. So you can `group_by(&:text)` and `v.map(&:file).uniq`